lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -424,164 +425,199 @@ ModulePass *llvm::createDxilCondenseResourcesPass() {
 INITIALIZE_PASS(DxilCondenseResources, "hlsl-dxil-condense", "DXIL Condense Resources", false, false)
 
 static
-bool LegalizeResourcesPHIs(Module &M, DxilValueCache *DVC) {
+bool EraseDeadBlocks(Module &M, DxilValueCache *DVC) {
+  std::unordered_set<BasicBlock *> Seen;
+  std::vector<BasicBlock *> WorkList;
+
+  bool Changed = false;
 
-  // Simple pass to collect resource PHI's
-  SmallVector<PHINode *, 8> PHIs;
   for (Function &F : M) {
-    for (BasicBlock &BB : F) {
-      for (Instruction &I : BB) {
-        if (PHINode *PN = dyn_cast<PHINode>(&I)) {
-          if (hlsl::dxilutil::IsHLSLObjectType(PN->getType())) {
-            PHIs.push_back(PN);
-          }
+    if (F.isDeclaration())
+      continue;
+    Seen.clear();
+    WorkList.clear();
+
+    auto Add = [&WorkList, &Seen](BasicBlock *BB) {
+      if (!Seen.count(BB)) {
+        WorkList.push_back(BB);
+        Seen.insert(BB);
+      }
+    };
+
+    Add(&F.getEntryBlock());
+
+    // Go through blocks
+    while (WorkList.size()) {
+      BasicBlock *BB = WorkList.back();
+      WorkList.pop_back();
+
+      if (BranchInst *Br = dyn_cast<BranchInst>(BB->getTerminator())) {
+        if (Br->isUnconditional()) {
+          BasicBlock *Succ = Br->getSuccessor(0);
+          Add(Succ);
         }
         else {
-          break;
+          bool IsConstant = false;
+          if (!DxilMDHelper::HasControlFlowHintToPreventFlatten(Br)) {
```

Comment:
Let's say it's a good thing to remove dead block, and just restore metadata when rewriting the br.